docs/how-to/model-debug.txt

```diff
@@ -0,0 +1,337 @@
+ Model Debugging in Determined
+###############################
+
+This document aims to provide useful guidelines for debugging models
+with Determined. Hopefully, it will help you become a power user of
+Determined, but if you get stuck please don't hesitate to contact us on
+TODO:ref-support-channels!
+
+This document focuses on model debugging, not cluster debugging, so it
+is assumed that you have already successfully :ref:`installed Determined
+<install-cluster>`.
+
+The basic steps for debugging are:
+
+   -  Model-related issues:
+
+        1. Does the original code run locally?
+        2. Does each method of your Trial class work locally?
+        3. Does local test mode work?
+
+   -  Docker- or Cluster-related issues
+
+        4. Does the original code run in a Notebook or Shell?
+        5. Does each method of your Trial class work in a Notebook or Shell?
+        6. Does local test mode work in a Notebook or Shell?
+
+   -  Higher-level issues
+
+        7. Does cluster test mode work with ``slots_per_trial=1``?
+        8. Does a single-GPU experiment work?
+        9. Does a multi-GPU experiment work?
+
+Model-related issues
+====================
+
+1. Does the original code run locally?
+--------------------------------------
+
+This step assumes you have have ported your model from some code outside
+of Determined. If your Trial code is not based on any such code, skip to
+Step 2.
+
+Probably you already know your code works, but this is a reminder to
+double-check before proceeding.
+
+2. Does each method of your Trial class work locally?
+-----------------------------------------------------
+
+This step assumes you have a working local environment for training. If
+you normally run your code in a docker environment of some instead, skip
+to Step 4.
+
+**How to test:** You should create some simple tests to verify that each
+method of your Trial is doing what you think it ought to be doing. There
+are some simple and practical examples of what these tests might look
+like for ``PyTorchTrial`` and ``TFKerasTrial`` in the
+:meth:`determined.TrainContext.from_config` documentation, but
+ultimately only you know what a reasonable test for your Trial needs to
+look like.
+
+**How to diagnose failures:** If you hit any issues running the methods
+of your Trial code locally, there are most likely errors in your Trial
+class, or possibly in the ``hyperparameters`` section of your config
+file. Ideally, breaking it down method-by-method like this makes finding
+and solving issues much faster.
+
+3. Does local test mode work?
+-----------------------------
+
+This step assumes you have a working local environment for training. If
+you do not, skip to Step 4.
+
+**How to test:** Simply create an experiment, passing ``--local --test``
+on the command line:
+
+.. code:: bash
+
+   det experiment create myconfig.yaml my_model_dir --local --test
+
+**How to diagnose failures:** Local test mode does very few things; it
```

Comment:
Non-blocking: I might move the explanation of what local test mode does up a bit higher, to the rationale for why you should do this debugging step.

I also feel like this discussion might be improved by a concrete example -- e.g., what kind of problem might be caught by local test mode that wouldn't be caught by the earlier debugging steps?

Reply:
Rewrote steps 1 and 2, let me know what you think.

Reply:
sg.